### ISSUE TITLE
Make `cargo --list` print description for custom subcommands

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/cargo#10662]
### Problem

Currently `cargo --list` prints descriptions only for the builtin subcommands. For third-party subcommands like `cargo expand` it does not print a description.

```console
$ cargo --list
Installed Commands:
    …
    doc                  Build a package's documentation
    expand
    fetch                Fetch dependencies of a package from the network
    …
```

### Proposed Solution

First of all, I don't want `cargo --list` to **run** any of the binaries (in order to, for example, run `--help` on them and parse the output, or some other handshake to cause the binary to print out a description of itself). Arbitrary binaries in the PATH are not necessarily safe to run. An arbitrary binary might completely ignore `--help` or any other handshake being passed by Cargo, and immediately do unwanted stuff instead, which would be unexpected and undesirable to the person running `cargo --list`.

Instead, Cargo subcommands should be able to embed a description into their compiled binary, which can then be directly retrieved from the binary by Cargo in a way that does not involve running it.

The way this is exposed to subcommands can be as simple as:

```rust
// subcommand's main.rs

cargo_subcommand_metadata::description!("…");

fn main() {
    /* … */
}
```

The underlying implementation would need to be platform-specific based on the file format used for executables on the platform. On platforms that use ELF executables, I would propose that this uses a Note. On those platforms the macro shown above would expand to:

```rust
#[used]
#[link_section = ".note.cargo.subcommand"]
static ELF_NOTE: ElfNote = {
    …
}
```

This note can then be retrieved efficiently from the ELF executable during `cargo --list` by using the [`object`](https://github.com/gimli-rs/object) crate to parse the executable's program header.

From testing on my machine with the `cargo-expand` crate, `cargo build --release` produces a 13 MB executable, and retrieving the subcommand description ELF note from it using the `object` crate takes 0.03 milliseconds, so this should be plenty fast enough for `cargo --list`.

We do not necessarily need to implement support for a large number of platforms immediately up front. Supporting just ELF would already benefit a large fraction of Cargo users.

### Notes

Yes, exactly.

**Comment by dtolnay** [comment]
As proof of concept, I have released a version of the `cargo-expand` crate containing a suitable ELF note (https://github.com/dtolnay/cargo-expand/pull/142) and created a draft PR #10663 which makes `cargo --list` retrieve the subcommand description from that ELF note.

![Screenshot from 2022-05-11 23-37-10](https://user-images.githubusercontent.com/1940490/168007135-64de96d6-21ee-4a1a-a921-2f13bd5ef7b7.png)

**Comment by epage** [comment]
Currently, cargo hard codes the descriptions for first-party cargo external subcommands like `cargo-clippy` and `cargo-fmt`.  This would also allow us to remove those hard coded descriptions and scale up better for adding more first-part external subcommands.  I've been toying with the idea of splitting some of cargo's built-in subcommands out as external subcommands.  This has the downside increasing the distribution size though it means we could possibly remove them in the minimal rustup profile.  On the other hand, modularization like this could help with a more approachable code base that is easier to review, dogfooding of cargo's APIs, and faster compile/test times for whichever section of cargo a developer is in.  This is blocked on our need to scale up our processes via workspaces which is blocked on the fact that we are part of the rust-lang/rust workspace and can't have our own.

**Comment by ehuss** [comment]
The Cargo Team discussed this a bit, and here are some questions or thoughts from that discussion:

* There were some concerns that the proposed solution seemed to be ELF-specific. Although this could be extended for other formats, like Windows, it seems risky to try to support that across all of Cargo's platforms. Additionally, some people use shell scripts or other kinds of executables that can't embed metadata.
* Would it be possible to use the `description` from `Cargo.toml` instead? I think there was some concern that what people put in `description` might be slightly different from what they would use for `cargo --list`, but I suspect in most cases it should be sufficient. If so, then one option that may be simpler is to embed the metadata in cargo's "install" JSON tracker.

**Comment by timmyjose** [comment]
I was thinking that I'd missed something in the documentation, but I see that this issue is still open. Hopefully it should be resolved sometime in the near future.

**Comment by epage** [comment]
Instead of doing a web request on every `cargo --list`, what if we stored the description inside of `crates2.json`?

**Comment by dtolnay** [comment]
Compared to putting descriptions into the executable, putting descriptions in .crates2.json has the disadvantage that it would only work for subcommands installed using `cargo install`. Subcommands installed using `cargo binstall` (https://github.com/cargo-bins/cargo-binstall/issues/1552), or just downloading a published release, would not be able to have descriptions in `cargo --list`.

**Comment by epage** [comment]
#13847 was proposing downloading descriptions from crates.io for plugins installed via `cargo install`, so definitely an improvement over that.

But yes, it doesn't cover other install cases.  Its also a lot lower barrier for adoption than accepting the embedded description scheme.

**Comment by Oakchris1955** [comment]
> Subcommands installed using `cargo binstall` (https://github.com/cargo-bins/cargo-binstall/issues/1552), or just downloading a published release, would not be able to have descriptions in `cargo --list`.

I think this is out of scope. But then again, one could alert the devs of `cargo-binstall` about the incoming change and have them implement a solution similar to ours

**Comment by weihanglo** [comment]
Web requests might lead to wrong information. People can have a script called `cargo-expand` under `$CARGO_HOME/bin`, but has nothing to do with [`cargo-expand`](https://crates.io/crates/cargo-expand) on crates.io.

To me, embedded description is more accurate an approach, and then description-in-crates2.json. Despite that, it depends on how accurate `cargo --list` need to be.

**Comment by Oakchris1955** [comment]
Perhaps the following logic could be implemented to solve most of the aforementioned issues:

1) When a user installs a subcommand, save the description in `.crates2.json`
2) When `cargo --list` is run:
    - For each subcommand, check if the relevant description field exists in `.crates2.json`
    - If not (because the subcommand would have installed before this issue got resolved/implemented), fetch the description from `crates.io` (if it was installed from there) and save it in `.crates2.json`
    - Show the description to the user

**Comment by epage** [comment]
> If not (because the subcommand would have installed before this issue got resolved/implemented), fetch the description from crates.io (if it was installed from there) and save it in .crates2.json

For myself, I don't see enough benefit to doing this.

**Comment by Oakchris1955** [comment]
I'd like to bring back attention to this issue. That is a relatively simple problem but with no agreed-upon solution. What can be done to go forward with this?

**Comment by weihanglo** [comment]
There are a couple of different solutions in this area, and each of them has pros and cons. Will need someone to get a summary of all approaches in comparison.

Also, I am not feeling good with expanding the `.crates2.json` file, particularly because it is a bit broken and not tracking the correct flags well. That said, it should have no compatibility issue if we end up adding a new field.

**Comment by Oakchris1955** [comment]
I see the following 2 approaches:
| Approach | Pro's | Con's | Notes/link |
|--------|--------|--------|--------|
| Obtain the description from either a `crates.io` API call or from the `Cargo.toml` of the program and storing it on `crates2.json` | Relatively easy to implement, should work with most commands installed through `cargo install` | Probably won't work with commands installed through third-party commands, like `cargo binstall`. Might cause issues if a user has a `cargo-COMMAND` in the `$PATH`. Furthermore, already-installed commands wouldn't display a description, unless if we were to fetch it for them during runtime and cache it then in `crates2.json` | <https://github.com/rust-lang/cargo/issues/10662#issuecomment-2100677121> |
| Embed the description into the resulting executable itself | Can be easily accessed pretty fast | Not all `cargo` custom commands will implement this, most of them probably won't. Cross-platform issues can also arise | <https://github.com/rust-lang/cargo/issues/10662#issue-1233490854> |

My point of view is biased, but I believe that the first approach would be simpler to implement and that it's disadvantages can be dealt with. As for the other solution, I believe that neither there is a reason to essentially duplicate the `description` field from `Cargo.toml` and embed it into an executable, where it may or may not exist depending on whether the developer knows or/and cares to implement such a feature for the custom `cargo` command. I believe that the majority of custom `cargo` commands published have at least a `description` field, so I suggest going with that

**Comment by weihanglo** [comment]
For why I said `.crates2.json` is a bit broken, see <https://github.com/rust-lang/cargo/issues/7322#issuecomment-1561463550>.
